---
title: "Examine_Clustered_TARGETdb_Fusions"
author: "bhaas"
date: '2022-09-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggrepel)
library(pheatmap)
library(cowplot)
library(ggforce)

```


```{r}

data = read.table(gzfile("../data/FusionInspector.v2.4.0.TARGET.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)


umap_data = read.table(gzfile("../data/FusionInspector.v2.4.0.TARGET.umap_coords.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)

# add TARGET db sample info

TARGETdb_sample_info = read.table("../data/TARGETdb.sample_info.tsv", sep="\t", header=T, stringsAsFactors = F )


TARGETdb_sample_info$project = str_replace(TARGETdb_sample_info$project, "TARGET-", "")
TARGETdb_sample_info$project = str_replace(TARGETdb_sample_info$project, "ALL-P[123]", "ALL")

#TARGETdb_sample_info  = TARGETdb_sample_info  %>% mutate(sample_type = paste0(project, "^", sample_type, "^", Cancer.Type) )

#TARGETdb_sample_info  = TARGETdb_sample_info  %>% mutate(sample_type = paste0(project, "^", sample_type) )

TARGETdb_sample_info  = TARGETdb_sample_info  %>% mutate(sample_type = paste0(project) )

data = left_join(data, TARGETdb_sample_info, by='sample')

data = data %>% mutate(tumor_or_normal = ifelse(grepl("Normal", sample_type), "normal", "tumor"))
```



```{r}

# so later we can examine reciprocal fusions

data = data %>% rowwise %>% mutate(partnerA = sort(str_split(fusion_name, "::")[[1]])[1])

data = data %>% rowwise() %>% mutate(partnerB = sort(str_split(fusion_name, "::")[[1]])[2])

data = data %>% rowwise() %>% mutate(sorted_fusion_name = paste0(partnerA, "::", partnerB) ) 

```



```{r}
# add the generic fusion annotations 

fusion_annotations = data %>% select(fusion_name, annots) %>% unique()

```


Below, we simplify annotations given by FusionAnnotator to a small number of categories that reflect the chromosomal positions of the fusion partners and what we might infer as structural rearrangements that would ultimately yield cis-spliced fusion transcripts.

The final 'structure_type' categories include:

- INTRACHROMOSOMAL
- INTERCHROMOSOMAL
- LOCAL_REARRANGEMENT (neighboring genes that wouldn't allow for simple readthrough explanations given the ref genome)
- READTHRU (short for readthrough cis-splicing of transcripts from neighboring genes)

```{r}
fusion_annotations = fusion_annotations  %>% mutate(annot_adj = annots)

fusion_annotations$annot_adj = str_replace(fusion_annotations$annot_adj, 
                                                      "NEIGHBORS_OVERLAP", "LOCAL_REARRANGEMENT")
fusion_annotations$annot_adj = str_replace(fusion_annotations$annot_adj, "LOCAL_INVERSION", "LOCAL_REARRANGEMENT") 


fusion_annotations$structure_type = "OTHER"

fusion_annotations = fusion_annotations%>% mutate(structure_type = ifelse(grepl("INTRACHROMOSOMAL", annot_adj),
                                                                               "IntraCHR", structure_type))

fusion_annotations = fusion_annotations %>% mutate(structure_type = ifelse(grepl("INTERCHROMOSOMAL", annot_adj), 
                                                                               "INTERCHR", structure_type))


fusion_annotations = fusion_annotations %>% mutate(structure_type = ifelse(grepl("LOCAL_REARRANGEMENT", annot_adj),
                                                                               "LOCAL_REARRANGEMENT", structure_type))

fusion_annotations = fusion_annotations %>% mutate(structure_type = ifelse(grepl("NEIGHBORS", annot_adj), 
                                                                               "READTHRU", structure_type))


message("Counts of structure types according to unique fusions among cosmic-like clusters")
table(fusion_annotations %>% select(fusion_name, structure_type) %>% unique() %>% pull(structure_type))

fusion_annotations = fusion_annotations %>% select(-annot_adj)


```




```{r}

all_cosmic_fusions = read.table("../data/cosmic.list", header=F, stringsAsFactors = F)[,1]

```


```{r}

data = left_join(data, umap_data %>% filter(umap_frac_data == 0.5), by=c('sample', 'fusion_name', 'LeftBreakpoint', 'RightBreakpoint'))


```



```{r}
# merge the tumor or normal samples per patient, selct the highest expressed fusion variant as representative

#samples_select = data %>% group_by(participant, tumor_or_normal) %>% filter(row_number() == 1) %>% pull(sample)
#data = data %>% filter(sample %in% samples_select)

nrow(data)

data = data %>% group_by(participant, fusion_name, LeftBreakpoint, RightBreakpoint, tumor_or_normal) %>%
  arrange(desc(FFPM)) %>%
  filter(row_number() == 1) %>%
  ungroup()

data = data %>% mutate(sample = paste0(participant, "^", tumor_or_normal))


# can restrict to only n primary fusion occurrences to examine how it impacts our findings.
# data = data %>% filter(is_primary) %>% group_by(fusion_name) %>% filter(row_number() <= 3) %>% ungroup()

nrow(data)
```




```{r}

umap.layout.cent = data %>% filter(! is.na(umap1)) %>% 
  select(umap1, umap2, leiden) %>% group_by(leiden) %>% summarize_all(mean)

```


```{r}

umap_plot = data %>% filter(! is.na(umap1)) %>%
  ggplot(aes(x=umap1, y=umap2)) + 
  geom_point(alpha = 0.3, size=0.05, aes(color=factor(leiden))) + theme_bw() +
  xlim(quantile(data$umap1, c(0.005, 0.995), na.rm = T)) +
  ylim(quantile(data$umap2, c(0.005, 0.995), na.rm=T))  +
  geom_label_repel(data=umap.layout.cent, aes(label = leiden), max.overlaps = Inf) + 
                             guides(colour = FALSE)


plot(umap_plot)
```




```{r}

# highlight the cosmic entries

cosmic_subset = data %>% filter(cosmic == TRUE & is_primary)

p = umap_plot + geom_point(data=cosmic_subset, color='black')

plot(p)


```


```{r}

ggsave(p, file="TARGETdb_umap_plot.pdf", width=12, height=10)


```




```{r}


max_cluster = max(data$leiden)

all_median_vals_df = NULL

data.scaled = data[, grepl("^adj", colnames(data))]


for (cluster_i in seq(max_cluster)) {

    #message("cluster: ", cluster_i)
    
    cluster_row_indices = which(data$leiden == cluster_i)
    
    feature_atts = data.scaled[cluster_row_indices,]
    
    median_vals = as_tibble(feature_atts) %>% gather() %>% group_by(key) %>% summarize(m=median(value))

    df = data.frame(median_vals$m, row.names=median_vals$key)
    df = as.data.frame(t(df))
    df$cluster = cluster_i

    all_median_vals_df = bind_rows(all_median_vals_df, df)

}


rownames(all_median_vals_df) = all_median_vals_df$cluster

all_median_vals_df = all_median_vals_df[,!c(colnames(all_median_vals_df)=='cluster')]

```



```{r}

median_vals_pca = prcomp(t(all_median_vals_df))
rows_d = dist(median_vals_pca$rotation[,1:6])
rowclust = hclust(as.dist(rows_d), "ward.D")


cols_d = dist(t(all_median_vals_df))

colclust = hclust(as.dist(cols_d), "ward.D")



p = pheatmap(all_median_vals_df, 
             cluster_rows=rowclust,
             cluster_cols=colclust,
             )


```

plot cosmic primary isoform fusion counts according to cluster

```{r}


cosmic_counts_per_cluster = cosmic_subset %>%
    group_by(fusion_name, leiden) %>% 
    tally()

cosmic_counts_per_cluster$leiden = factor(cosmic_counts_per_cluster$leiden, levels=rev(p$tree_row$labels[p$tree_row$order]))

empty_counts_df = data.frame(fusion_name="NA", leiden=factor(seq(max_cluster), levels=rev(p$tree_row$labels[p$tree_row$order])), n=0)

cosmic_counts_per_cluster = bind_rows(empty_counts_df, cosmic_counts_per_cluster)



# plot it


cosmic_primary_fusion_counts_per_cluster_plot = cosmic_counts_per_cluster %>% ggplot(aes(x=leiden, y=n, fill=fusion_name)) + geom_col() + coord_flip() + ggtitle("COSMIC primary isoform count")


plot(cosmic_primary_fusion_counts_per_cluster_plot)
```


```{r}

# without colors

cosmic_sum_counts_per_cluster = cosmic_counts_per_cluster %>% group_by(leiden) %>% summarize(sum_n = sum(n))


cosmic_sum_counts_per_cluster_plot = cosmic_sum_counts_per_cluster %>% ggplot(aes(x=leiden, y=sum_n)) + geom_col() + coord_flip() + ggtitle("COSMIC primary isoform count")



plot(cosmic_sum_counts_per_cluster_plot)


```




```{r}

# all fusions, top isoforms


top_fusion_counts_per_cluster = data %>% 
    filter(is_primary) %>%
    select(fusion_name, leiden) %>% 
    group_by(fusion_name, leiden) %>% 
    tally()

top_fusion_counts_per_cluster$leiden = factor(top_fusion_counts_per_cluster$leiden, levels=rev(p$tree_row$labels[p$tree_row$order]))

empty_counts_df = data.frame(fusion_name="NA", leiden=factor(seq(max_cluster), levels=rev(p$tree_row$labels[p$tree_row$order])), n=0)


top_fusion_counts_per_cluster = bind_rows(empty_counts_df, top_fusion_counts_per_cluster)

sum_fusion_counts_per_cluster = top_fusion_counts_per_cluster %>% group_by(leiden) %>% summarize(sum_n = sum(n))

sum_fusion_counts_per_cluster_plot = sum_fusion_counts_per_cluster %>% ggplot(aes(x=leiden, y=sum_n)) + geom_col() + coord_flip() + ggtitle("Fusion primary isoform count")

plot(sum_fusion_counts_per_cluster_plot)
```

plot cosmic enrichment per cluster

```{r}

cosmic_sum_counts_per_cluster$fusion_type = "cosmic"
sum_fusion_counts_per_cluster$fusion_type = "all"
cosmic_and_all_counts_df = bind_rows(cosmic_sum_counts_per_cluster, sum_fusion_counts_per_cluster)

cosmic_and_all_counts_df = cosmic_and_all_counts_df %>% spread(key=fusion_type, value=sum_n)
cosmic_and_all_counts_df = cosmic_and_all_counts_df %>% mutate(frac_cosmic = cosmic/all)

cosmic_frac_enrichment_plot = cosmic_and_all_counts_df %>% ggplot(aes(x=leiden, y=frac_cosmic)) + 
    geom_col() + coord_flip() + ggtitle("Fraction COSMIC fusions")
plot(cosmic_frac_enrichment_plot)
```


Create single summary plot

```{r}


cosmic_sum_counts_per_cluster_plot_NoLegend = cosmic_sum_counts_per_cluster_plot + 
    theme(legend.position = "none")

pg = plot_grid(p[[4]], cosmic_sum_counts_per_cluster_plot_NoLegend, sum_fusion_counts_per_cluster_plot, cosmic_frac_enrichment_plot, 
          nrow=1, rel_widths = c(3,1,1,1) )

plot(pg)

```


```{r}

ggsave(pg, file="TARGETdb_heatmap_collage_plot.pdf", width=20, height=13)

```


```{r}

frac_cluster_predictions = data %>% filter(is_primary) %>% group_by(leiden, fusion_cluster_att) %>% tally() %>%
  mutate(pct=prop.table(n)) 

#empty_counts_df = data.frame(leiden=factor(seq(max_cluster), levels=rev(p$tree_row$labels[p$tree_row$order])), pct=0, fusion_cluster_att=NA)

#frac_cluster_predictions = bind_rows(empty_counts_df, frac_cluster_predictions)


frac_cluster_predictions$leiden = factor(frac_cluster_predictions$leiden, levels=rev(p$tree_row$labels[p$tree_row$order]))

frac_cluster_att_predictions_plot = frac_cluster_predictions  %>%
  mutate(pct = ifelse(is.na(fusion_cluster_att ), 0, pct)) %>%
  ggplot(aes(x=leiden, y=pct, fill=fusion_cluster_att)) + geom_col() + coord_flip() 


plot(frac_cluster_att_predictions_plot)

```

```{r}

cosmic_sum_counts_per_cluster_plot_NoLegend = cosmic_sum_counts_per_cluster_plot + 
    theme(legend.position = "none")

pg = plot_grid(p[[4]], cosmic_sum_counts_per_cluster_plot_NoLegend, sum_fusion_counts_per_cluster_plot, cosmic_frac_enrichment_plot, 
               frac_cluster_att_predictions_plot + theme(legend.position = "none"),
          nrow=1, rel_widths = c(3,1,1,1,1) )

plot(pg)



```

# Examine COSMIC enrichment stat significance

```{r}

# combine into single table and examine fraction cosmic fusions per cluster

cosmic_counts_by_cluster = full_join(cosmic_sum_counts_per_cluster %>% select(leiden, sum_cosmic = sum_n), 
                                     sum_fusion_counts_per_cluster %>% select(leiden, sum_fusions = sum_n) , by='leiden')


total_cosmic = cosmic_counts_by_cluster %>% summarize(tot_cosmic = sum(sum_cosmic), tot_fusions = sum(sum_fusions))

cosmic_counts_by_cluster = cosmic_counts_by_cluster  %>% mutate(frac_cosmic = sum_cosmic / sum_fusions)

cosmic_counts_by_cluster = bind_cols(cosmic_counts_by_cluster, total_cosmic)

cosmic_counts_by_cluster %>% arrange(desc(sum_cosmic))

```


```{r}

# statistical significance of COSMIC fusion enrichment




run_fishers_exact_test = function(cosmic_count, fusion_count, tot_cosmic_fusions, total_fusions) {
    
    message("cosmic_count: ", cosmic_count)
    message("fusion count: ", fusion_count)
  
    
    non_cosmic_count = fusion_count - cosmic_count
    tot_noncosmic_fusions  = total_fusions - tot_cosmic_fusions
    
    confusion_matrix = matrix(c(cosmic_count, non_cosmic_count,
                                tot_cosmic_fusions - cosmic_count, tot_noncosmic_fusions - non_cosmic_count),
                              nrow=2, byrow = T)    
    print(confusion_matrix)
    
    f = fisher.test(confusion_matrix, alternative = 'greater')
    
    return(f$p.value)
    
}


cosmic_counts_by_cluster = cosmic_counts_by_cluster  %>% rowwise() %>% mutate(
  pval = run_fishers_exact_test(
         sum_cosmic, sum_fusions, tot_cosmic, tot_fusions)
)  %>% mutate(negLOGp = -1 * log10(pval))


cosmic_counts_by_cluster %>% arrange(desc(negLOGp))


```

```{r}
cosmic_counts_by_cluster %>% ggplot(aes(x=leiden, y=negLOGp)) + geom_col() +
  ggtitle("TARGET COSMIC fusion enrichment per predicted clusters") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))



```

```{r}
cosmic_counts_by_cluster$pval_BH = p.adjust(cosmic_counts_by_cluster$pval, method="BH")

cosmic_counts_by_cluster %>% arrange(pval_BH)

# cluster 21, adj pval = 3.7e-151
```

# Examine COSMIC and COSMIC-like fusions in cluster 21

```{r}

COSMIC_cluster = "21"


data %>% filter(cosmic & is_primary) %>% group_by(fusion_name, sample_type) %>%  tally() %>%
  ggplot(aes(x=fusion_name, y=n, fill=sample_type)) + geom_col() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))

```


```{r}
# examine cosmic fusions according to the clusters they fall in

data %>% filter(is_primary & cosmic) %>% group_by(fusion_name, leiden) %>% tally() %>%
  ggplot(aes(x=fusion_name, y=n, fill=factor(leiden))) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))



```



```{r}

fusion_sample_type_counts = data %>% filter(is_primary) %>% group_by(fusion_name, leiden, project, sample_type) %>% tally(name='fusion_sample_type_counts')

sample_type_counts = data %>% select(sample, sample_type) %>% unique() %>% group_by(sample_type) %>% tally(name='sample_type_counts')

fusion_sample_type_counts = left_join(fusion_sample_type_counts, sample_type_counts, by='sample_type')


fusion_sample_type_counts = fusion_sample_type_counts %>% mutate(sample_type_frac = fusion_sample_type_counts / sample_type_counts)

fusion_sample_type_counts = fusion_sample_type_counts %>% mutate(cosmic = (fusion_name %in% all_cosmic_fusions))


```



```{r}

# examine cosmic fusions according to sample type representation

fusion_sample_type_counts %>% filter(cosmic) %>% ggplot(aes(x=fusion_name, y=sample_type_frac, fill=sample_type)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```

```{r}
# list the most prevalent cosmic fusions by sample type
fusion_sum_sample_type_fractions = fusion_sample_type_counts %>% 
  select(fusion_name, project, sample_type_frac, sample_type) %>% 
  group_by(fusion_name, project, sample_type) %>% 
  summarize(sum_frac = sum(sample_type_frac)) %>% arrange(desc(sum_frac))
```

```{r}
fusion_sum_sample_type_fractions %>% filter(fusion_name %in% all_cosmic_fusions) 

```












############################################
# Explore new potential fusions of interest
############################################

- exclude those found in normals here
- rank by tumor sample fractions and proportion in C21


```{r}

fusions_found_in_normals = data %>% filter(tumor_or_normal == "normal") %>% select(fusion_name) %>% unique() %>% pull(fusion_name)

length(fusions_found_in_normals)
# 1036
```

```{r}

fusion_fraction_C21 = data %>% filter(is_primary) %>%  
  group_by(fusion_name, sorted_fusion_name, leiden) %>% 
  tally(name='count_in_C21') %>% 
  mutate(frac_leiden_C21=prop.table(count_in_C21)) %>% 
  filter(leiden==COSMIC_cluster)

fusion_fraction_C21
```
```{r}

# examine fraction of COSMIC fusion instances found in C21

fusion_fraction_C21  = fusion_fraction_C21 %>% mutate(cosmic = (fusion_name %in% all_cosmic_fusions) )

fusion_fraction_C21 %>% filter(cosmic) %>% arrange(frac_leiden_C21)

# min of 10% here

```

```{r}

# get total fusion abundance across TARGET samples
fusion_all_sample_counts = data %>% filter(is_primary) %>% group_by(fusion_name) %>% tally(name='tot_sample_count')

fusion_fraction_C21 = left_join(fusion_fraction_C21, fusion_all_sample_counts, by='fusion_name')


fusion_fraction_C21 
```





```{r}

# combine with earlier tcga/gtex occurrence stats
tcga_gtex_fusion_stats = read.table(gzfile("../data/TCGA_n_GTEx.STAR-Fusion.v1.7.fusion_preds.stats.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)


fusion_fraction_C21  = left_join(fusion_fraction_C21, tcga_gtex_fusion_stats, by='fusion_name')

fusion_fraction_C21
```

```{r}

# examine the cosmic-here set wrt our earlier tcga/gtex findings

fusion_fraction_C21 %>% filter(cosmic) %>% arrange(desc(count_in_C21))


```


```{r}

write.table(fusion_fraction_C21 %>% filter(cosmic) %>% arrange(desc(count_in_C21)), file="C21_cosmic_entries.tsv", quote=F, sep="\t", row.names = F)

```


# select top new candidates

```{r}

# find some new candidates of interest

top_new_candidates = fusion_fraction_C21 %>%  
  filter(! fusion_name %in% fusions_found_in_normals) %>%
  filter(tot_sample_count >= 3) %>%
  #filter(frac_leiden_C21 >= 0.10) %>%
  filter(normal == 0 | is.na(normal)) %>%
  arrange(desc(tot_sample_count))


#top_new_candidates = top_new_candidates %>% filter(! grepl("NEIGHBORS", annots))


top_new_candidates
```



```{r}

# incorporate the sample type fractions

top_new_candidates = left_join(top_new_candidates, fusion_sum_sample_type_fractions,  by='fusion_name', )

top_new_candidates 
```



```{r}

# include annotations
top_new_candidates = left_join(top_new_candidates, fusion_annotations, by='fusion_name')



top_new_candidates 


```



```{r}
sum_sum_frac = top_new_candidates %>% select(fusion_name, tot_sample_count, cosmic, sum_frac, project, sample_type) %>% unique() %>%
  group_by(fusion_name) %>% mutate(sum_sum_frac = sum(sum_frac)) %>% select(fusion_name, sum_sum_frac) %>% unique()

top_new_candidates = left_join(top_new_candidates, sum_sum_frac, by='fusion_name')
 
```


```{r}

candidates_sample_frac_plot = top_new_candidates  %>% 
  group_by(fusion_name, project) %>% arrange(desc(sum_frac)) %>% filter(row_number()==1) %>% ungroup() %>%
  ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=sum_frac, fill=project)) + geom_col() +  
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

candidates_sample_frac_plot 

```


```{r}

# indicate if in cosmic or not

cosmic_entries_plot = top_new_candidates %>% select(fusion_name, sum_sum_frac, cosmic) %>% unique() %>%
  mutate(cosmic = ifelse(cosmic, TRUE, NA)) %>% 
  ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=TRUE, fill=cosmic)) + 
  geom_tile() +
  theme_bw() +
            theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
            theme(axis.title.y=element_blank(), axis.text.y = element_blank(), axis.ticks.y=element_blank()) +
            scale_fill_manual(na.value="white", values="purple") +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

cosmic_entries_plot 



```



```{r}

# examine reciprocal translocations for C21


# define fusion gene pairs found in C21 samples
C21_fusion_containing_samples_n_sorted_fusions =  data %>% filter(leiden==COSMIC_cluster) %>% select(sample, sorted_fusion_name) %>% unique() 


# define the reciprocal fusions among these
# - restricted to C21 samples
# - the fusion_name A::B and B::A must show up in the same sample
reciprocal_fusions_data = left_join( data, C21_fusion_containing_samples_n_sorted_fusions, by=c('sample', 'sorted_fusion_name')) %>%
       filter(is_primary) %>%
       group_by(sample, fusion_name, sorted_fusion_name) %>% unique() %>% group_by(sample, sorted_fusion_name) %>% tally() %>% filter(n>1) %>% ungroup()

# capture both A::B and B::A occurrences of fusions for these samples.
reciprocal_fusions_data = left_join(reciprocal_fusions_data, 
                                    data, 
                                    by=c('sample', 'sorted_fusion_name'))



# now further restrict to those fusions among our top candidates as earlier defined.
top_new_candidate_sorted_fusions = top_new_candidates %>% select(sorted_fusion_name) %>% unique() %>% pull(sorted_fusion_name)

reciprocal_fusions_data = reciprocal_fusions_data %>% filter(sorted_fusion_name %in% top_new_candidate_sorted_fusions)

reciprocal_fusions_data %>% filter(is_primary) %>% arrange(sample, sorted_fusion_name)




```

```{r}


top_new_candidates = left_join(top_new_candidates, 
                               reciprocal_fusions_data %>% filter(leiden == COSMIC_cluster) %>% select(fusion_name) %>% 
                                 unique() %>% mutate(reciprocal_fusion = TRUE),
                               by='fusion_name')

top_new_candidates

```

```{r}

reciprocal_fusion_plot = top_new_candidates %>% select(fusion_name, sum_sum_frac, reciprocal_fusion) %>% unique() %>%
  ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=TRUE, fill=reciprocal_fusion)) + 
  geom_tile() +
  theme_bw() +
            theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
            theme(axis.title.y=element_blank(), axis.text.y = element_blank(), axis.ticks.y=element_blank()) +
            scale_fill_manual(na.value="white", values="blue")  +
   theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

reciprocal_fusion_plot

```


```{r}


# examine fusion sample correlation matrix for only the reciprocal fusion subset

#fusion_reciprocal_sample_C21_info = data %>% 
#  filter(is_primary & 
#           sorted_fusion_name %in% reciprocal_fusions_data$sorted_fusion_name & 
#           sorted_fusion_name %in% top_new_candidates$sorted_fusion_name) %>% 
#  select(fusion_name, project, sample, sorted_fusion_name) %>% unique() 

#write.table(fusion_reciprocal_sample_C21_info %>% select(project, fusion_name) %>% unique(), 
#  file="C21_project_fusion.reciprocal.info.tsv", quote=F, sep="\t", col.names=F, row.names=F)


write.table(reciprocal_fusions_data %>% select(project, fusion_name) %>% unique(), 
  file="C21_project_fusion.reciprocal.info.tsv", quote=F, sep="\t", col.names=F, row.names=F)



write.table(reciprocal_fusions_data %>% select(fusion_name, sample) %>%  mutate(found=1) %>% unique() %>% spread(key=sample, value=found, fill=0),
            file="C21_project_fusion.reciprocal.sample_matrix.tsv", quote=F, sep="\t", row.names=F)


fusion_sample_matrix = read.table("C21_project_fusion.reciprocal.sample_matrix.tsv", row.names=1, header=T, sep="\t")
write.table(t(fusion_sample_matrix), file="C21_project_fusion.reciprocal.sample_matrix.tsv", quote=F, sep="\t", row.names=T)


# Make the sample correlation plot like so using PtR:
# trinityrnaseq/Analysis/DifferentialExpression/PtR -m C21_project_fusion.reciprocal.sample_matrix.tsv -s C21_project_fusion.reciprocal.info.tsv --heatmap --sample_cor_matrix --img_width 20 --img_height 20


```





# Want COSMIC-like cluster stats and frame change stats too.



```{r}

top_new_candidate_fusions = top_new_candidates %>% select(fusion_name) %>% unique() %>% pull(fusion_name)

```

```{r}

cosmic_like_pred_data = data %>% filter(fusion_name %in% top_new_candidate_fusions) %>% filter(is_primary) %>% 
  mutate(fusion_cluster_att = ifelse(is.na(fusion_cluster_att), "Other", fusion_cluster_att)) %>% 
  group_by(fusion_name, fusion_cluster_att) %>% tally() %>% mutate(frac_fusions=prop.table(n))

cosmic_like_pred_data
```


```{r}


cosmic_like_pred_data = bind_rows(cosmic_like_pred_data %>% select(fusion_name, fusion_cluster_att, frac_fusions),
          top_new_candidates %>% select(fusion_name, frac_fusions =frac_leiden_C21) %>% unique() %>% mutate(fusion_cluster_att = "frac_in_C21") )

cosmic_like_pred_data 

```



```{r}


cosmic_like_pred_data$fusion_cluster_att = factor(cosmic_like_pred_data$fusion_cluster_att, levels=rev(c('Other', 'COSMIC-like', 'expr_microH_RT_artifact?', 'frac_in_C21')))


cosmic_like_pred_plot = left_join(cosmic_like_pred_data, sum_sum_frac) %>% 
  ggplot(aes(x=fusion_name, y=fusion_cluster_att, fill=frac_fusions)) +
        geom_tile() +
        theme_bw() +
        theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
        ylab("Cluster Type") +
        theme(axis.text.y = element_text(size=rel(0.7)))  +
        scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white") +
         theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

cosmic_like_pred_plot

```


```{r}

fusion_coding_effects = read.table(gzfile("../data/fusions.coding_effects.abridged.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)


```


```{r}

candidate_fusion_coding_data = left_join(data %>% filter(is_primary) %>% filter(fusion_name %in% top_new_candidate_fusions),
                                         fusion_coding_effects,
                                         by=c('fusion_name', 'LeftBreakpoint', 'RightBreakpoint'))

candidate_fusion_coding_data 
```

```{r}

candidate_fusion_coding_frac_data = candidate_fusion_coding_data %>% select(fusion_name, PROT_FUSION_TYPE, sample) %>% 
  mutate(PROT_FUSION_TYPE = ifelse(PROT_FUSION_TYPE==".", "Other", PROT_FUSION_TYPE)) %>% 
  group_by(fusion_name, PROT_FUSION_TYPE) %>% tally() %>% mutate(frac_fusions=prop.table(n)) 

candidate_fusion_coding_frac_plot = left_join(candidate_fusion_coding_frac_data, sum_sum_frac) %>%  
  ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=PROT_FUSION_TYPE, fill=frac_fusions)) +
        geom_tile() +
        theme_bw() +
        theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
        ylab("Cluster Type") +
        theme(axis.text.y = element_text(size=rel(0.7)))  +
        scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white")  +
         theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))


candidate_fusion_coding_frac_plot
```




```{r}
# chromosome structure

chrom_structure_plot = left_join(top_new_candidates, sum_sum_frac) %>% select(fusion_name, structure_type, sum_sum_frac)  %>% unique() %>%
    ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=0, fill=structure_type)) +
  geom_tile() +
   theme_bw() +
       theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(),
          axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank()
          ) +
 
         theme(panel.border = element_blank(), panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

chrom_structure_plot

```




```{r}
# combine plots into a single figure


pg = plot_grid(cosmic_like_pred_plot,
               candidate_fusion_coding_frac_plot,
               chrom_structure_plot,
               cosmic_entries_plot,
               reciprocal_fusion_plot,
               candidates_sample_frac_plot,
                 ncol=1,
                 align='v',
                 axis='lr',
                 rel_heights = c(0.2, 0.2, 0.15, 0.15, 0.15,0.70)
                 )


plot(pg)


```

```{r}

ggsave(pg, file="TARGET_summary_plot.pdf", height=8, width=14)


```

```{r}

# Make the short list:

top_new_candidates_top_sample_frac = top_new_candidates %>% group_by(fusion_name) %>% 
  arrange(desc(sum_frac)) %>%
  filter(row_number()==1) %>%
  ungroup() %>%
  select(fusion_name, tot_sample_count, sum_frac, frac_leiden_C21, cosmic, normal, tumor, logTN, project, sample_type, annots, structure_type, sorted_fusion_name) %>%
  arrange(desc(sum_frac))


top_new_candidates_top_sample_frac

```



```{r}
write.table(top_new_candidates_top_sample_frac, file="TARGET.top_new_candidates.tsv", quote=F, sep="\t", row.names = F)

```



```{r}


blood_each_sample_type_heatmap = top_new_candidates %>% 

 ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=sample_type)) +
        geom_tile(aes(fill=sum_frac)) +
        facet_col(vars(project), , scales = "free_y", space = "free") +

         theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
        ylab("sample_type") +
        xlab("fusion") +
        theme_bw() +
        theme(axis.text.y = element_text(size=rel(1)))  +
        scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white") +
        scale_x_discrete(expand = c(0, 0)) +    
        scale_y_discrete(expand = c(0, 0)) +  
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

plot(blood_each_sample_type_heatmap)

```

```{r}
ggsave(blood_each_sample_type_heatmap, file="blood_each_sample_type_heatmap.pdf", width=15, height=8)

```


```{r}
blood_top_sample_type_heatmap = top_new_candidates %>% 
  group_by(fusion_name, project) %>% arrange(desc(sum_frac)) %>% filter(row_number() == 1) %>% ungroup() %>%

 ggplot(aes(x=reorder(fusion_name, -sum_sum_frac), y=reorder(project, desc(project)))) +
        geom_tile(aes(fill=sum_frac)) +

#         theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())
#+
        ylab("sample_type") +
        xlab("fusion") +
        theme_bw() +
#        theme(axis.text.y = element_text(size=rel(1)))  +
        scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white") +
       theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
       theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black"))


plot(blood_top_sample_type_heatmap )
```




```{r}


pg = plot_grid(cosmic_like_pred_plot,
               candidate_fusion_coding_frac_plot,
               chrom_structure_plot,
               cosmic_entries_plot,
               reciprocal_fusion_plot,
               blood_top_sample_type_heatmap ,
                 ncol=1,
                 align='v',
                 axis='lr',
                 rel_heights = c(0.2, 0.2, 0.15, 0.1, 0.1, 0.70)
                 )


plot(pg)

ggsave(pg, file="TARGET_fusions_of_interest.pdf", width=18, height=12)




```


```{r}

fusion_sample_type_counts %>% filter(fusion_name == "TCF3::PBX1")


```





