---
title: "TARGET fusion analysis"
author: "bhaas"
date: '2022-08-16'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```


```{r}

target_sample_info = read.table("../data/TARGETdb.sample_info.tsv", sep="\t", header=T, stringsAsFactors = F)

target_sample_info = target_sample_info %>% mutate(tumor_or_normal = ifelse(grepl("Normal", sample_type), "normal", "tumor"))

target_sample_info$project = str_replace(target_sample_info$project, "TARGET-", "")
target_sample_info$project = str_replace(target_sample_info$project, "ALL-P[\\d]", "ALL")

target_project_sample_counts = target_sample_info %>% select(sample, project, tumor_or_normal) %>% unique() %>% 
  group_by(project, tumor_or_normal) %>%  tally(name='total_sample_count') %>% arrange(project, desc(total_sample_count))


target_project_sample_counts

#  Acute Lymphoblastic Leukemia (ALL, 666 tumors), Acute Myeloid Leukemia (AML, 217 tumors and 20 normal samples), Clear Cell Sarcoma of the Kidney (CCSK, 13 tumors), Neuroblastoma (NB,  157 tumors), Osteosarcoma (OS, 88 tumors), Rhabdoid Tumor (RT, 63 tumors and 6 normal), and Wilms Tumor (WT, 130 tumors and 6 normal).

```

```{r}

target_sample_info %>% nrow()
# 1366 samples

target_sample_info %>% select(participant) %>% unique() %>% nrow()
# 1230 participants

```

```{r}
target_project_participant_counts = target_sample_info %>% select(participant, project, tumor_or_normal) %>% unique() %>% 
  group_by(project, tumor_or_normal) %>%  tally(name='total_participant_count') %>% arrange(project, desc(total_participant_count))


target_project_participant_counts
```




# STARF results

```{r}

STARF_data = read.table(gzfile("../data/STAR-Fusion.v1.10.0.TARGETdb.tsv.gz"),
                           header=T, sep="\t", stringsAsFactors = F)


STARF_data$fusion_name = str_replace(STARF_data$fusion_name, "--", "::")

## some initial cleanup
STARF_data = STARF_data %>% 
  filter(! grepl("chrM", annots)) %>%
  filter(! grepl("HLA", fusion_name)) %>%
  filter(! (grepl("^IG[HKVL]", LeftGene) & grepl("^IG[HKVL]", RightGene)))


# set primary isoform
STARF_data = STARF_data %>% group_by(sample, fusion_name) %>% arrange(desc(FFPM)) %>% mutate(is_primary = (row_number()==1) ) %>% ungroup()


# add sample info
STARF_data = left_join(STARF_data, target_sample_info, by='sample')

cosmic_fusions = read.table("../data/cosmic.list")[,1]
STARF_data = STARF_data %>% mutate(cosmic = (fusion_name %in% cosmic_fusions))


```




```{r}

# how many STAR-F fusions variants
STARF_data %>% select(sample, fusion_name) %>% nrow()
# 228612

# how many STAR-F fusion occurrences
STARF_data %>% filter(is_primary) %>% select(sample, fusion_name)  %>% nrow()
# 213786

# number of unique gene pairs:
STARF_data %>% select(fusion_name) %>% unique() %>% nrow()
# 108661

```



# FI results

```{r}

# input data

FI_data = read.table(gzfile("../data/FusionInspector.v2.4.0.TARGET.tsv.gz"), sep="\t", com='', header=T, stringsAsFactors = F)

FI_data = FI_data %>% 
  filter(! grepl("chrM", annots)) %>%
  filter(! grepl("HLA", fusion_name)) %>%
  filter(! (grepl("^IG[HKVL]", LeftGene) & grepl("^IG[HKVL]", RightGene)))


FI_data = left_join(FI_data, target_sample_info, by='sample')

```


```{r}


# how many FI fusions variants
FI_data %>% select(sample, fusion_name) %>% nrow()
# 217614

# how many FI fusion occurrences

FI_data %>% filter(is_primary) %>% select(sample, fusion_name)  %>% nrow()
# 182807

```


182807 / 213786 = 0.86

so, 86% of the STAR-F identified fusions were further 'in silico validated' by FusionInspector.


```{r}

# count number of distinct fusion gene pairs represented:
FI_data %>% select(fusion_name)  %>% unique() %>% nrow()

# 88307

```


# COSMIC fusions in TARGET w/ comparison to TCGA

```{r}



TARGET_cosmic_fusion_participant_counts = STARF_data %>% filter(cosmic & is_primary) %>% select(participant, project, tumor_or_normal, fusion_name) %>% unique() %>% group_by(project, tumor_or_normal, fusion_name) %>% tally(name='participant_count_w_fusion') %>% ungroup()

TARGET_cosmic_fusion_participant_counts = left_join(TARGET_cosmic_fusion_participant_counts, target_project_participant_counts, by=c('project', 'tumor_or_normal'))


TARGET_cosmic_fusion_participant_counts = TARGET_cosmic_fusion_participant_counts %>% 
  mutate(sample_type_frac = participant_count_w_fusion / total_participant_count)


fusion_name_sum_participant_counts = TARGET_cosmic_fusion_participant_counts %>% group_by(fusion_name) %>% summarize(sum_participants = sum(participant_count_w_fusion)) %>% ungroup()

TARGET_cosmic_fusion_participant_counts = left_join(TARGET_cosmic_fusion_participant_counts, fusion_name_sum_participant_counts, by='fusion_name')

TARGET_cosmic_fusion_participant_counts %>% arrange(desc(sample_type_frac))
```


```{r}

# how many unique COSMIC fusions found in TARGET
TARGET_cosmic_fusion_participant_counts %>% select(fusion_name) %>% unique() %>% nrow()

# 30

```


```{r}


TARGET_cosmic_participant_count_plot = TARGET_cosmic_fusion_participant_counts %>% ggplot(aes(x=reorder(fusion_name, -sum_participants), y=participant_count_w_fusion, fill=project)) + geom_col() +
   theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
   theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black"))

TARGET_cosmic_participant_count_plot

```


```{r}


TARGET_cosmic_proj_frac_plot = TARGET_cosmic_fusion_participant_counts %>% ggplot(aes(x=reorder(fusion_name, -sum_participants), y=sample_type_frac, fill=project)) + geom_col() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
   theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black"))

TARGET_cosmic_proj_frac_plot
```





```{r}

# examine our earlier TCGA / COSMIC results:

tcga_gtex_stats = read.table(gzfile("../data/TCGA_n_GTEx.STAR-Fusion.v1.7.fusion_preds.stats.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)

tcga_gtex_stats_cosmic = tcga_gtex_stats %>% filter(fusion_name %in% cosmic_fusions) 

cosmic_fusion_participant_counts = bind_rows(TARGET_cosmic_fusion_participant_counts %>% select(project, fusion_name, participant_count_w_fusion), 
                                        
                                        tcga_gtex_stats_cosmic %>% filter(tumor>0) %>% 
                                          mutate(project="TCGA") %>% 
                                          select(project, fusion_name, sample_count_w_fusion = tumor)
                                        ) 

```


```{r}

cosmic_fusion_participant_counts = cosmic_fusion_participant_counts %>% mutate(in_TCGA = (fusion_name %in% tcga_gtex_stats$fusion_name)) 

```


```{r}
cosmic_fusion_participant_counts  %>% 
  filter(fusion_name %in% TARGET_cosmic_fusion_participant_counts$fusion_name) %>%
  ggplot(aes(x=reorder(fusion_name, in_TCGA), y=project, fill=log(participant_count_w_fusion+1))) + geom_tile() + 
  theme_bw() +
   scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white") +
       theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
       theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black"))

                                        
```
                             
```{r}

# which cosmic fusions are unique to TARGET set as compared to TCGA

cosmic_fusion_participant_counts %>% filter(! fusion_name %in% tcga_gtex_stats$fusion_name) %>%
  ggplot(aes(x=fusion_name, y=project, fill=log(participant_count_w_fusion+1))) + geom_tile() + 
  
   theme_bw() +
   scale_fill_continuous(high = "#132B43", low = "#56B1F7", na.value="white") +
       theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
       theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black")) +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))
                                        


```


```{r}
# how many cosmic are unique to TARGET - not found in TCGA?

cosmic_fusion_participant_counts %>% ungroup() %>% filter(! fusion_name %in% tcga_gtex_stats$fusion_name) %>% 
  select(fusion_name) %>% unique() %>% nrow()

# 22

```

```{r}

TCGA_fusions =  tcga_gtex_stats %>% filter(tumor > 0) %>% pull(fusion_name)

```


```{r}
in_TCGA_indicator_plot = TARGET_cosmic_fusion_participant_counts  %>% mutate(in_TCGA = ifelse(fusion_name %in% TCGA_fusions, TRUE, NA ) ) %>%
  ggplot(aes(x=reorder(fusion_name, -sum_participants), y='in_TCGA', fill=in_TCGA)) + geom_tile() + 
  theme_bw() +
       theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
       theme(panel.border = element_blank(), 
             panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(), 
             axis.line = element_line(colour = "black")) +
   scale_fill_manual(na.value="white", values="purple")

 in_TCGA_indicator_plot   
 
 # 8 shown in TCGA
 
```

Figure S7a


```{r}

# joint plot for supp fig.

library(cowplot)

plot_grid(TARGET_cosmic_participant_count_plot + theme(axis.title.x=element_blank(), axis.text.x=element_blank()),  
          TARGET_cosmic_proj_frac_plot + theme(axis.title.x=element_blank(), axis.text.x=element_blank()),
           in_TCGA_indicator_plot ,
          ncol=1,
          align='v',
          axis='lr',
          rel_heights = c(0.25, 0.25, 0.1)
          )



```



# What are the fusions found in the TARGET normal samples?

```{r}

normal_FI_fusions = left_join(FI_data %>% filter(tumor_or_normal == "normal"),
                                                 tcga_gtex_stats,
                                                 by='fusion_name'
                                                 )


```


```{r}
# count up the normal TARGET fusions

# how many normal samples w fusions:
normal_FI_fusions  %>% select(sample) %>% unique() %>% nrow()
# 32 samples.

normal_FI_fusions %>% filter(is_primary) %>% nrow()
# 1865 normal fusion occurrences


normal_FI_fusions %>% filter(is_primary) %>% select(fusion_name) %>% unique() %>% nrow()
# 1036 normal fusion gene pairs

```


```{r}

# examine overlap with GTEx normals
normal_FI_fusions %>% filter(normal > 0) %>% select(fusion_name) %>% unique()
# 428 normal fusion gene pairs overlap with GTEx (41% of unique fusion gene pairs), and so more than half (59%) of these normal fusions are unique to the small set of normals here.

```

```{r}

normal_FI_fusions = normal_FI_fusions %>% mutate(in_GTEx = (normal > 0 | ! is.na(normal)))

```

```{r}
# examine the most commonly encountered normal fusions in TARGET

normal_FI_fusions %>% filter(is_primary) %>% group_by(fusion_name, in_GTEx) %>% tally() %>% arrange(desc(n))


```


```{r}

# how many of these normal fusions are recurrent among TARGET normals, and of these, how many are in GTEx?
normal_FI_fusions %>% filter(is_primary) %>% group_by(fusion_name, in_GTEx) %>% tally() %>% filter(n>1) %>% nrow()
# 199

# 199/1036 = 19% of the TARGET normal fusions are recurrent.

# how many of these 199 are in GTEx?
normal_FI_fusions %>% filter(is_primary & in_GTEx) %>% group_by(fusion_name, in_GTEx) %>% tally() %>% filter(n>1) %>% nrow()
# 179

# so 179/199 = 90% of the recurrent normal TARGET fusions are also found in GTEx.




```




